src/mgmt/srv_query.c

```diff
@@ -0,0 +1,176 @@
+ */
+#define D_LOGFAC	DD_FAC(mgmt)
+
+#include <daos_srv/bio.h>
+#include <daos_srv/smd.h>
+
+#include "srv_internal.h"
+
+static void
+bio_health_query(void *arg)
+{
+	struct mgmt_bio_health	*mbh = arg;
+	struct dss_module_info	*info = dss_get_module_info();
+	struct bio_xs_context	*bxc;
+	struct bio_dev_state	*health_stats;
+
+	D_ASSERT(info != NULL);
+	D_DEBUG(DB_MGMT, "BIO health stats query on xs:%d, tgt:%d\n",
+		info->dmi_xs_id, info->dmi_tgt_id);
+
+	bxc = info->dmi_nvme_ctxt;
+	if (bxc == NULL) {
+		D_ERROR("BIO NVMe context not initialized for xs:%d, tgt:%d\n",
+			info->dmi_xs_id, info->dmi_tgt_id);
+		goto out;
+	}
+
+	health_stats = bio_get_dev_state(bxc);
+	if (health_stats == NULL)
+		D_ERROR("Error getting BIO device state\n");
+
+	uuid_copy(mbh->mb_devid, mbh->mb_dev_info->sdi_id);
+	mbh->mb_error_count = health_stats->bds_error_count;
+	mbh->mb_media_errors = health_stats->bds_media_errors;
+	mbh->mb_read_errs = health_stats->bds_bio_read_errs;
+	mbh->mb_write_errs = health_stats->bds_bio_write_errs;
+	mbh->mb_unmap_errs = health_stats->bds_bio_unmap_errs;
+	mbh->mb_temperature = health_stats->bds_temperature;
+	mbh->mb_temp = health_stats->bds_temp_warning ? true : false;
+	mbh->mb_spare = health_stats->bds_avail_spare_warning ? true : false;
+	mbh->mb_readonly = health_stats->bds_read_only_warning ? true : false;
+	mbh->mb_reliability = health_stats->bds_dev_reliabilty_warning ? true : false;
+	mbh->mb_volatile = health_stats->bds_volatile_mem_warning ? true : false;
+
+out:
+	smd_free_dev_info(mbh->mb_dev_info);
+}
+
+int
+ds_mgmt_bio_health_query(struct mgmt_bio_health *mbh, uuid_t dev_uuid,
+			char *tgt)
+{
+	ABT_thread		 thread;
+	int			 tgt_id;
+	int			 rc = 0;
+
+	if (uuid_is_null(dev_uuid) && strlen(tgt) == 0) {
+		/* Either dev uuid or tgt id needs to be specified for query */
+		D_ERROR("Neither dev_uuid or tgt_id specified for BIO query\n");
+		return -DER_INVAL;
+	}
+
+	/*
+	 * Query per-server metadata (SMD) to get either target ID(s) for given
+	 * device or alternatively the device mapped to a given target.
+	 */
+	if (!uuid_is_null(dev_uuid)) {
+		rc = smd_dev_get_by_id(dev_uuid, &mbh->mb_dev_info);
+		if (rc != 0) {
+			D_ERROR("Device UUID:"DF_UUID" not found\n",
+				DP_UUID(dev_uuid));
+			return rc;
+		}
+		if (mbh->mb_dev_info->sdi_tgts == NULL) {
+			D_ERROR("No targets mapped to device\n");
+			return -DER_NONEXIST;
+		}
+		/* Default tgt_id is the first mapped tgt */
+		tgt_id = mbh->mb_dev_info->sdi_tgts[0];
+	} else {
+		tgt_id = atoi(tgt);
+		rc = smd_dev_get_by_tgt(tgt_id, &mbh->mb_dev_info);
+		if (rc != 0) {
+			D_ERROR("Tgt_id:%d not found\n", tgt_id);
+			return rc;
+		}
+		uuid_copy(dev_uuid, mbh->mb_dev_info->sdi_id);
+	}
+
+	D_DEBUG(DB_MGMT, "Querying BIO Health Data for dev:"DF_UUID"\n",
+		DP_UUID(dev_uuid));
+
+	/* Create a ULT on the tgt_id */
+	D_DEBUG(DB_MGMT, "Starting ULT on tgt_id:%d\n", tgt_id);
+	rc = dss_ult_create(bio_health_query, mbh, DSS_ULT_AGGREGATE,
```

Comment:
I just chose this so it would hit the case to return DSS_MAIN_XS_ID(tgt_id). I believe this is correct? I can add a new DSS_ULT_HEALTH tag also. Or should it be DSS_ULT_MISC and return (DSS_MAIN_XS_ID + 1 % DSS_XS_NR_TOTAL)?

Reply:
No, I don't think you need to create a new tag. I think we can cleanup the dss ult types a bit in  the future.

Reply:
ok maybe DSS_ULT_MISC would be more appropriate then